bonus/mtl.md

```diff
@@ -0,0 +1,515 @@
+instance Monad m => MonadReader r (ReaderT r m) where
+  ask' = error "ask' -- ReaderT r m"
+  asks' = error "asks' -- ReaderT r m"
+  local' = error "local' -- ReaderT r m"
+
+instance Monad m => MonadError e (ExceptT e m) where
+  throwError' = error "throwError' -- ExceptT e m"
+  catchError' = error "catchError' -- ExceptT e m"
+
+-- These two instances lift the functionality of one
+-- transformer through another.
+instance MonadReader r m => MonadReader r (ExceptT e m) where
+  ask' = error "ask' -- lift through ExceptT"
+  asks' = error "asks' -- lift through ExceptT"
+  local' = error "local' -- lift through ExceptT"
+
+instance MonadError e m => MonadError e (ReaderT r m) where
+  throwError' = error "throwError' -- lift through ReaderT"
+  catchError' = error "catchError' -- lift through ReaderT"
+```
+
+<details>
+<summary>Solution</summary>
+
+```haskell ignore
+instance Monad m => MonadReader r (ReaderT r m) where
+  ask' = ask
+  asks' = asks
+  local' = local
+
+instance Monad m => MonadError e (ExceptT e m) where
+  throwError' = throwError
+  catchError' = catchError
+
+instance MonadReader r m => MonadReader r (ExceptT e m) where
+  ask' = lift ask'
+  asks' = lift . asks'
+  local' = mapExceptT . local'
+
+instance MonadError e m => MonadError e (ReaderT r m) where
+  throwError' = lift . throwError'
+  catchError' = liftCatch catchError'
+```
+
+</details>
+
+
+## Writing all the Instances Stinks!
+
+Yes. This is called the "O(n^2) instances" problem. Each bundle of
+monad operations wants its own class, instances of that class for one
+or more transformers (if you have multiple transformers providing
+different implementations), plus instances that lift through every
+other possible transformer.
+
+This is a lot of boilerplate, and it's important to know two things:
+
+1. It's not always possible to lift one instance through
+   another. Such instances do not and should not exist.
+
+2. Monad transformers, in general, do not commute. `StateT s (ExceptT
+   e m)` is a different beast to `ExceptT e (StateT s m)` - the former
+   throws away the state on error, while the latter preserves it.
+
+Algebraic effect systems attempt to cut down boilerplate, but make
+compromises of their own. There is active research in this area.
+
+
+## `AppM` Revisited
+
+The payoff of all this setup (helpfully encapsulated in the `mtl`
+library, which builds atop `transformers`) is that we can specify our
+application monad as a stack of transformers atop `IO`, and use GHC's
+`GeneralizedNewtypeDeriving` extension to give us all the instances
+immediately:
+
+```haskell
+data Env -- Dummy type for the sake of example
+
+-- This is the Level07 AppM, which also passes around an environment
+newtype AppMtl e a = AppMtl (ReaderT Env (ExceptT e IO) a)
+  deriving (Functor, Applicative, Monad, MonadError e, MonadReader Env)
+
+runAppMtl :: AppMtl e a -> Env -> IO (Either e a)
+runAppMtl (AppMtl m) env = runExceptT $ runReaderT m env
```

Comment:
Maybe make it a stronger callout to the fact that we don't have to eat this turd salad because others have done the annoying bit, so for almost every usecase, you can derive thusly and crack on with the interesting things...